cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java

```diff
@@ -200,21 +187,54 @@ private void closeDataset(DatasetCacheKey key, Dataset dataset) {
   }
 
   @Override
+  public void dismissDataset(Dataset dataset) {
+    Preconditions.checkNotNull(dataset);
+    // static datasets cannot be dismissed
+    if (staticDatasets.containsValue(dataset)) {
+      LOG.warn("Attempt to dismiss static dataset {} from dataset cache", dataset);
+      return;
+    }
+    if (txContext == null || !(dataset instanceof TransactionAware)) {
+      dismissSafely(dataset);
+    } else {
+      // it is a tx-aware: it may participate in a transaction, so mark it as to be dismissed after the tx:
+      // the transaction context will call dismissSafely() for this dataset when the tx is complete.
+      txContext.dismissAfterTx((TransactionAware) dataset);
+    }
+    // remove from activeTxAwares in any case - a dismissed dataset does not need to participate in external tx
+    // iterates over all datasets but we do not expect this map to become large
+    for (Map.Entry<DatasetCacheKey, TransactionAware> entry : activeTxAwares.entrySet()) {
```

Comment:
Also, if I try that, IntelliJ warns me about a suspicious method call... Since the complexity is the same (requires iterating over the map), I find it ok to leave the explicit for loop here.